packages/babel-helper-module-transforms/src/dynamic-import.ts

```diff
@@ -1,9 +1,67 @@
 // Heavily inspired by
 // https://github.com/airbnb/babel-plugin-dynamic-import-node/blob/master/src/utils.js
 
+import type { File } from "@babel/core";
 import * as t from "@babel/types";
 import template from "@babel/template";
 
+// TODO(Babel 8): Remove this
+const DynamicImportSpecifierHelper = template.expression.ast`
+function (s) {
+  try {
+    return Promise.resolve("" + s);
+  } catch(e) {
+    return Promise.reject(e);
+  }
+}`;
+// @ts-expect-error undocumented _compact node property
+DynamicImportSpecifierHelper._compact = true;
+
+export function buildDynamicImportSpecifier(
+  file: File,
+  node: t.CallExpression,
+  pureDynamicImportSpecifier: boolean,
+): {
+  pure: boolean;
+  specifier: t.Expression;
+} {
+  const [specifier] = node.arguments;
+
+  if (
+    t.isStringLiteral(specifier) ||
+    (t.isTemplateLiteral(specifier) && specifier.quasis.length === 0)
+  ) {
+    return { pure: true, specifier };
+  }
+
+  if (pureDynamicImportSpecifier) {
+    return {
+      pure: true,
+      specifier: template.expression.ast`\`\${${specifier}}\``,
+    };
+  }
+
+  let helper: t.Identifier;
+  if (file.availableHelper("dynamicImportSpecifier")) {
+    helper = file.addHelper("dynamicImportSpecifier");
+  } else {
+    // Fallback for @babel/helpers <= 7.20.6, manually add helper function
+    // TODO(Babel 8): Remove this
+    helper = file.get("fallbackDynamicImportSpecifier");
+    if (!helper) {
+      const id = file.scope.generateUidIdentifier("defineAccessor");
+      file.scope.push({
+        id,
+        init: DynamicImportSpecifierHelper,
+      });
+      file.set("fallbackDynamicImportSpecifier", (helper = id));
+    }
+    helper = t.cloneNode(helper);
+  }
+
+  return { pure: false, specifier: t.callExpression(helper, [specifier]) };
+}
+
 export function getDynamicImportSource(
   node: t.CallExpression,
 ): t.StringLiteral | t.TemplateLiteral {
```

Comment:
Could the old function `getDynamicImportSource` be deleted safely (won't cause compatibility issue)?

Reply:
It is a public export, so better keep it until Babel 8.